### ISSUE TITLE
[VB]False-positive "Cast is redundant" for `DirectCast`

### ISSUE THREAD

**Issue by tats-u** [dotnet/roslyn#72505]
**Version Used**: 

main (20 Jan 2024) in sharplab

**Steps to Reproduce**:

1. Write the following code:

https://sharplab.io/#v2:EYLgtghglgdgNAGxAN2AHwPIAcAuUD2MABAJIwBmApgE5EbnkCwAUCyWFvtTgM5EDKATx45KYFgAUArsARQAxkQDCCCDz5KWRbUWmyFAmUQCyACnxE1dYACtK8nAEotO1wBEoYIlCIBBPmSiAOY0RAC8RB7U9jhKajjmcKQwwTTOzK7unkR8/pH4MgiU4SYQOAAWAHT8AI7cplDpmTpKhDz4RZUA6tRQogAysJSmPE06AKIwACaGwCyTMypqPEA=

```vbnet
Option Infer Off

Imports System
Public Class C
    Public Sub M(o as Object)
        Dim i As Integer = DirectCast(o, Integer)
        Dim s As Double = Math.Sqrt(i)
        Console.WriteLine(s)
    End Sub
End Class
```

A minimal repro, with source-code provided, is ideal.  Using [sharplab](https://sharplab.io/) is preferred for compiler/language issues whenever possible.

2. Run "Refactor: Remove Unnecessary Cast" as the diagnostic suggests

**Diagnostic Id**:

If this is a report about a bug in an analyzer, please include the diagnostic if possible (e.g. `"IDE0030"`).

IDE0004 Cast is redundant.

**Expected Behavior**:

No diagnostics even without `Option Infer Off`

**Actual Behavior**:

IDE0004 appears at `DirectCast`

`DirectCast(o, Integer)` is changed to `o`, whose behavior is different especially when the content in `o` is other than `Integer`

`DirectCast` is different from `CType` when the source is `Object`.

**Comment by CyrusNajmabadi** [comment]
Closing out as we are not investing in this area.
